cppcache/integration/test/CachingProxyTest.cpp

```diff
@@ -0,0 +1,162 @@
+
+#include <geode/Cache.hpp>
+#include <geode/CacheFactory.hpp>
+#include <geode/RegionFactory.hpp>
+#include <geode/RegionShortcut.hpp>
+#include <geode/Serializable.hpp>
+
+#include "framework/Cluster.h"
+
+using apache::geode::client::Cacheable;
+using apache::geode::client::CacheableString;
+using apache::geode::client::CacheFactory;
+using apache::geode::client::RegionShortcut;
+using apache::geode::client::Serializable;
+
+namespace CachingProxy {
+
+class CachingProxyTest : public ::testing::Test {
+ protected:
+  CachingProxyTest() {
+    cluster.start();
+
+    cluster.getGfsh()
+        .create()
+        .region()
+        .withName("region")
+        .withType("PARTITION")
+        .execute();
+  }
+
+  ~CachingProxyTest() override = default;
+
+  void SetUp() override {}
+
+  void TearDown() override {}
+
+  Cluster cluster = Cluster{LocatorCount{1}, ServerCount{1}};
+  std::string key = std::string("scharles");
+  std::string value = std::string("Sylvia Charles");
+};
+
+TEST_F(CachingProxyTest, LocalRemoveAfterLocalInvalidate) {
+  auto cache = CacheFactory().create();
+
+  cache.getPoolManager()
+      .createFactory()
+      .addLocator("localhost", cluster.getLocatorPort())
+      .create("pool");
+
+  auto region = cache.createRegionFactory(RegionShortcut::CACHING_PROXY)
+                    .setPoolName("pool")
+                    .create("region");
+
+  region->put(key, value);
+
+  auto user = region->get(key);
+
+  region->localInvalidate(key);
+
+  bool resultLocalRemove = region->localRemove(key, value);
+  ASSERT_FALSE(resultLocalRemove);
+
+  resultLocalRemove = region->localRemove(
+      key, static_cast<std::shared_ptr<Cacheable>>(nullptr));
+  ASSERT_TRUE(resultLocalRemove);
+  cache.close();
+}
+
+TEST_F(CachingProxyTest, RemoveAfterInvalidate) {
+  auto cache = CacheFactory().create();
+
+  cache.getPoolManager()
+      .createFactory()
+      .addLocator("localhost", cluster.getLocatorPort())
+      .create("pool");
+
+  auto region = cache.createRegionFactory(RegionShortcut::CACHING_PROXY)
+                    .setPoolName("pool")
+                    .create("region");
+
+  region->put(key, value);
+
+  region->invalidate(key);
+
+  bool resultRemove = region->remove(key, value);
+  ASSERT_FALSE(resultRemove);
+
+  resultRemove =
+      region->remove(key, static_cast<std::shared_ptr<Cacheable>>(nullptr));
+  ASSERT_TRUE(resultRemove);
+
+  cache.close();
+}
+
+TEST_F(CachingProxyTest, RemoveAfterLocalInvalidate) {
+  auto cache = CacheFactory().create();
+
+  cache.getPoolManager()
+      .createFactory()
+      .addLocator("localhost", cluster.getLocatorPort())
+      .create("pool");
+
+  auto region = cache.createRegionFactory(RegionShortcut::CACHING_PROXY)
+                    .setPoolName("pool")
+                    .create("region");
+
+  region->put(key, value);
+
+  region->localInvalidate(key);
+
+  bool resultRemove =
+      region->remove(key, static_cast<std::shared_ptr<Cacheable>>(nullptr));
```

Comment:
We don't have a specialization for `nullptr`? We should probably add that.

Reply:
Nice idea. Added a function for this in CacheableBuiltins.hpp so don't need to cast:
     region->remove(key, nullptr);